test/tests/functional/pbs_hook_modifyvnode_state_changes.py

```diff
@@ -0,0 +1,512 @@
+            e.reject("state sets: bad v sets")
+        if len(v_o.extract_state_strs()) == len(v_o.extract_state_ints()):
+            pbs.logmsg(pbs.LOG_DEBUG, "state sets: good v_o sets")
+        else:
+            e.reject("state sets: bad v_o sets")
+        e.accept()
+    except SystemExit:
+        pass
+    except:
+        pbs.event().reject("%s hook failed with %s" % (
+                           pbs.event().hook_name, sys.exc_info()[:2]))
+    """
+    hook_body = textwrap.dedent(hook_body)
+    return hook_body
+
+
+def get_hook_body_reverse_node_state():
+    hook_body = """
+    import pbs
+    e = pbs.event()
+    pbs.logmsg(pbs.LOG_DEBUG, "pbs.__file__:" + pbs.__file__)
+    # this is backwards as it's a reverse lookup.
+    for value, key in pbs.REVERSE_NODE_STATE.items():
+        pbs.logmsg(pbs.LOG_DEBUG, "key:%s value:%s" % (key, value))
+    e.accept()
+    """
+    hook_body = textwrap.dedent(hook_body)
+    return hook_body
+
+
+class TestPbsModifyvnodeStateChanges(TestFunctional):
+
+    """
+    Test the modifyvnode hook by inducing various vnode state changes and
+    inspecting the pbs log for expected values.
+    """
+
+    def setUp(self):
+        TestFunctional.setUp(self)
+        Job.dflt_attributes[ATTR_k] = 'oe'
+
+        self.server.cleanup_jobs()
+
+    def tearDown(self):
+        TestFunctional.tearDown(self)
+        # Delete managers and operators if added
+        attrib = ['operators', 'managers']
+        self.server.manager(MGR_CMD_UNSET, SERVER, attrib)
+
+    def checkLog(self, start_time, mom, check_up, check_down):
+        self.server.log_match("set_vnode_state;vnode.state=",
+                              starttime=start_time)
+        self.server.log_match("show_vnode_state;name=",
+                              starttime=start_time)
+        self.server.log_match("name: v=", starttime=start_time)
+        self.server.log_match("state: v=", starttime=start_time)
+        self.server.log_match("last_state_change_time: v=",
+                              starttime=start_time)
+        self.server.log_match("good times", starttime=start_time)
+        self.server.log_match("good names", starttime=start_time)
+        self.server.log_match("good states", starttime=start_time)
+        self.server.log_match("good v sets", starttime=start_time)
+        self.server.log_match("good v_o sets", starttime=start_time)
+        if check_up:
+            self.server.log_match("Node;%s;node up" % mom,
+                                  starttime=start_time)
+        if check_down:
+            self.server.log_match("Node;%s;node down" % mom,
+                                  starttime=start_time)
+
+    @tags('smoke')
+    def test_hook_state_changes_00(self):
+        """
+        Test: induce a variety of vnode state changes with debug turned on
+        and inspect the pbs log for expected entries
+        """
+        self.logger.info("---- %s TEST STARTED ----" % get_method_name(self))
```

Comment:
Because I use the "TEST STARTED" and "TEST ENDED" strings in my own work, unless there are objections I prefer leaving them in. Note I changed all the test logging from info to debug, including these.